beacon-chain/sync/data_columns_reconstruct.go

```diff
@@ -0,0 +1,187 @@
+		}
+
+		recoveredBlobs = append(recoveredBlobs, recoveredBlob)
+		log.WithFields(logrus.Fields{
+			"elapsed": time.Since(start),
+			"index":   blobIndex,
+			"root":    fmt.Sprintf("%x", blockRoot),
+		}).Debug("Recovered blob")
+	}
+
+	return recoveredBlobs, nil
+}
+
+// getSignedBlock retrieves the signed block corresponding to the given root.
+// If the block is not available, it waits for it.
+func (s *Service) getSignedBlock(
+	ctx context.Context,
+	blockRoot [fieldparams.RootLength]byte,
+) (interfaces.ReadOnlySignedBeaconBlock, error) {
+	blocksChannel := make(chan *feed.Event, 1)
+	blockSub := s.cfg.blockNotifier.BlockFeed().Subscribe(blocksChannel)
+	defer blockSub.Unsubscribe()
+
+	// Get the signedBlock corresponding to this root.
+	signedBlock, err := s.cfg.beaconDB.Block(ctx, blockRoot)
+	if err != nil {
+		return nil, errors.Wrap(err, "block")
+	}
+
+	// If the block is here, return it.
+	if signedBlock != nil {
+		return signedBlock, nil
+	}
+
+	// Wait for the block to be available.
+	for {
+		select {
+		case blockEvent := <-blocksChannel:
+			// Check the type of the event.
+			data, ok := blockEvent.Data.(*statefeed.BlockProcessedData)
+			if !ok || data == nil {
+				continue
+			}
+
+			// Check if the block is the one we are looking for.
+			if data.BlockRoot != blockRoot {
+				continue
+			}
+
+			// This is the block we are looking for.
+			return data.SignedBlock, nil
+		case err := <-blockSub.Err():
+			return nil, errors.Wrap(err, "block subscriber error")
+		case <-ctx.Done():
+			return nil, errors.New("context canceled")
+		}
+	}
+}
+
+func (s *Service) reconstructDataColumns(ctx context.Context, verifiedRODataColumn blocks.VerifiedRODataColumn) error {
+	// Lock to prevent concurrent reconstruction.
+	s.dataColumsnReconstructionLock.Lock()
+	defer s.dataColumsnReconstructionLock.Unlock()
+
+	// Get the block root.
+	blockRoot := verifiedRODataColumn.BlockRoot()
+
+	// Get the columns we store.
+	storedColumnsIndices, err := s.cfg.blobStorage.ColumnIndices(blockRoot)
+	if err != nil {
+		return errors.Wrap(err, "columns indices")
+	}
+
+	storedColumnsCount := len(storedColumnsIndices)
+	numberOfColumns := fieldparams.NumberOfColumns
+
+	// If less than half of the columns are stored, reconstruction is not possible.
+	// If all columns are stored, no need to reconstruct.
+	if storedColumnsCount < numberOfColumns/2 || storedColumnsCount == numberOfColumns {
+		return nil
+	}
+
+	// Load the data columns sidecars.
+	dataColumnSideCars := make([]*ethpb.DataColumnSidecar, 0, storedColumnsCount)
+	for index := range storedColumnsIndices {
+		dataColumnSidecar, err := s.cfg.blobStorage.GetColumn(blockRoot, index)
+		if err != nil {
+			return errors.Wrap(err, "get column")
+		}
+
+		dataColumnSideCars = append(dataColumnSideCars, dataColumnSidecar)
+	}
+
+	// Recover blobs.
+	recoveredBlobs, err := recoverBlobs(dataColumnSideCars, storedColumnsCount, blockRoot)
```

Comment:
Its very expensive to go from columns -> blobs and then back to columns. Once we recover all the cells, we can simply use that to reseed the missing columns.

Reply:
Actually we don't really do `columns -> blobs -> columns` but `(partial) dataColumnsSidecars -> blobs -> (complete) dataColumnsSidecars`.

To get the complete dataColumnsSidecars, we need the KZG proofs, which are computed by blobs, using `cKzg4844.ComputeCellsAndKZGProofs(blob)`.

Do you see a way to compute these proofs without going trough blobs?

Reply:
The KZG Proofs are computed for all blobs no ? for each column sidecar. You can simply reuse the previous commitments here from the earlier columns.